torch/csrc/DeviceSpec.cpp

```diff
@@ -0,0 +1,176 @@
+{
+  std::ostringstream oss;
+  oss << "DeviceSpec(device_type=\'" << deviceTypeString(self->device_type) << "\'";
+  if (!self->is_default) {
+    oss << ", device_index=" << self->device_index;
+  }
+  oss << ")";
+  return THPUtils_packString(oss.str().c_str());
+}
+
+PyObject *THPDeviceSpec_str(THPDeviceSpec *self)
+{
+  std::ostringstream oss;
+  if (!self->is_default) {
+    oss << deviceTypeString(self->device_type) << ":" << self->device_index;
+  } else {
+    oss << deviceTypeString(self->device_type);
+  }
+  return THPUtils_packString(oss.str().c_str());
+}
+
+PyObject *THPDeviceSpec_pynew(PyTypeObject *type, PyObject *args, PyObject *kwargs)
+{
+  HANDLE_TH_ERRORS
+  static torch::PythonArgParser parser({
+    "DeviceSpec(Device device)",
+    "DeviceSpec(String device_type, int64_t? device_index=-1)"
+  });
+  torch::ParsedArgs<2> parsed_args;
+  auto r = parser.parse(args, kwargs, parsed_args);
+  if (r.idx == 0) {
+    auto device = r.device(0);
+    return THPDeviceSpec_New(device.device_type, device.device_index, device.is_default);
+  } else if (r.idx == 1) {
+    auto as_device = r.device(0);  // this works, because device can take strings
+    auto device_type = r.string(0);
+    if (!as_device.is_default) {
+      throw std::runtime_error("device_type must not include index because index is passed explicitly " + device_type);
+    }
+
+    auto is_default = r.isNone(1);
+    auto device_index = r.toInt64WithDefault(1, -1);
+    // make sure this is constructible
+    auto device = torch::Device(as_device.device_type, device_index, is_default);
+    return THPDeviceSpec_New(device.device_type, device.device_index, device.is_default);
+  }
+  Py_RETURN_NONE;
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject *THPDeviceSpec_device_type(THPDeviceSpec *self)
+{
+  HANDLE_TH_ERRORS
+  return THPUtils_packString(deviceTypeString(self->device_type).c_str());
+  Py_RETURN_NONE;
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject *THPDeviceSpec_device_index(THPDeviceSpec *self)
+{
+  HANDLE_TH_ERRORS
+  if (self->is_default) {
+    Py_RETURN_NONE;
+  } else {
+    return THPUtils_packInt64(self->device_index);
+  }
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject *THPDeviceSpec_cuda_device_index(THPDeviceSpec *self)
+{
+  HANDLE_TH_ERRORS
+  if (self->device_type == torch::DeviceType::CUDA) {
+    return THPUtils_packInt64(self->device_index);
+  }
+  std::ostringstream oss;
+  oss << "cuda_device_index only supported on cuda device, got: " << deviceTypeString(self->device_type);
+  throw std::runtime_error(oss.str());
+  END_HANDLE_TH_ERRORS
+}
+
+
+typedef PyObject *(*getter)(PyObject *, void *);
+
+static struct PyGetSetDef THPDeviceSpec_properties[] = {
+  {"device_type",       (getter)THPDeviceSpec_device_type, nullptr, nullptr, nullptr},
+  {"device_index",      (getter)THPDeviceSpec_device_index, nullptr, nullptr, nullptr},
+  {"cuda_device_index", (getter)THPDeviceSpec_cuda_device_index, nullptr, nullptr, nullptr},
```

Comment:
nit: the `device_` prefix might be unnecessary. It's clear that this object describes a device, so there's no need to be overly verbose

Reply:
device_type / device_index are the names that TensorFlow uses; I'm fine changing them though.

Reply:
I mean... I can easily imagine myself doing this:
```python
device = torch.device('cuda:4')
...
device.device_index # why?!
```
That just looks unnecessary. I agree it's nice to conform to existing APIs, but meh.